presto-hive/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java

```diff
@@ -0,0 +1,340 @@
+        result.setTableType(table.getTableType());
+        result.setParameters(table.getParameters());
+        result.setPartitionKeys(table.getPartitionColumns().stream().map(MetastoreUtil::toMetastoreApiFieldSchema).collect(Collectors.toList()));
+        result.setSd(makeStorageDescriptor(table.getTableName(), table.getDataColumns(), table.getStorage()));
+        result.setPrivileges(privileges);
+        result.setViewOriginalText(table.getViewOriginalText().orElse(null));
+        result.setViewExpandedText(table.getViewExpandedText().orElse(null));
+        return result;
+    }
+
+    public static org.apache.hadoop.hive.metastore.api.Partition toMetastoreApiPartition(Partition partition)
+    {
+        org.apache.hadoop.hive.metastore.api.Partition result = new org.apache.hadoop.hive.metastore.api.Partition();
+        result.setDbName(partition.getDatabaseName());
+        result.setTableName(partition.getTableName());
+        result.setValues(partition.getValues());
+        result.setSd(makeStorageDescriptor(partition.getTableName(), partition.getColumns(), partition.getStorage()));
+        result.setParameters(partition.getParameters());
+        return result;
+    }
+
+    public static Table fromMetastoreApiTable(org.apache.hadoop.hive.metastore.api.Table table)
+    {
+        StorageDescriptor storageDescriptor = table.getSd();
+        if (storageDescriptor == null) {
+            throw new PrestoException(HIVE_INVALID_METADATA, "Table is missing storage descriptor");
+        }
+
+        Table.Builder tableBuilder = new Table.Builder()
+                .setDatabaseName(table.getDbName())
+                .setTableName(table.getTableName())
+                .setOwner(table.getOwner())
+                .setTableType(table.getTableType())
+                .setDataColumns(storageDescriptor.getCols().stream()
+                        .map(MetastoreUtil::fromMetastoreApiFieldSchema)
+                        .collect(Collectors.toList()))
+                .setPartitionColumns(table.getPartitionKeys().stream()
+                        .map(MetastoreUtil::fromMetastoreApiFieldSchema)
+                        .collect(Collectors.toList()))
+                .setParameters(table.getParameters() == null ? ImmutableMap.of() : table.getParameters())
+                .setViewOriginalText(makeOptional(table.getViewOriginalText()))
+                .setViewExpandedText(makeOptional(table.getViewExpandedText()));
+
+        fromMetastoreApiStorageDescriptor(storageDescriptor, tableBuilder.getStorageBuilder(), table.getTableName());
+
+        return tableBuilder.build();
+    }
+
+    public static Partition fromMetastoreApiPartition(org.apache.hadoop.hive.metastore.api.Partition partition)
+    {
+        StorageDescriptor storageDescriptor = partition.getSd();
+        if (storageDescriptor == null) {
+            throw new PrestoException(HIVE_INVALID_METADATA, "Partition is missing storage descriptor");
```

Comment:
`throw new PrestoException(HIVE_INVALID_METADATA, "Partition does not contain a storage descriptor: " + partition)`